vulnerabilities/pipelines/remove_ghost_packages.py

```diff
@@ -0,0 +1,84 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# VulnerableCode is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/vulnerablecode for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import logging
+
+from aboutcode.pipeline import LoopProgress
+from fetchcode.package_versions import SUPPORTED_ECOSYSTEMS
+from fetchcode.package_versions import versions
+from packageurl import PackageURL
+from univers.version_range import RANGE_CLASS_BY_SCHEMES
+
+from vulnerabilities.models import Package
+from vulnerabilities.pipelines import VulnerableCodePipeline
+
+
+class RemoveGhostPackagePipeline(VulnerableCodePipeline):
+    """Detect and remove packages that do not exist upstream."""
+
+    @classmethod
+    def steps(cls):
+        return (cls.remove_ghost_packages,)
+
+    def remove_ghost_packages(self):
+        detect_and_remove_ghost_packages(logger=self.log)
+
+
+def detect_and_remove_ghost_packages(logger=None):
+    """Use fetchcode to validate the package indeed exists upstream."""
+    interesting_packages_qs = (
+        Package.objects.filter(type__in=SUPPORTED_ECOSYSTEMS)
+        .filter(qualifiers="")
+        .filter(subpath="")
+    )
+
+    distinct_packages = interesting_packages_qs.values("type", "namespace", "name").distinct(
+        "type", "namespace", "name"
+    )
+
+    distinct_packages_count = distinct_packages.count()
+    package_iterator = distinct_packages.iterator(chunk_size=2000)
+    progress = LoopProgress(total_iterations=distinct_packages_count, logger=logger)
+
+    ghost_package_count = 0
+
+    for package in progress.iter(package_iterator):
+        ghost_package_count += remove_ghost_package(
+            package=package,
+            interesting_packages_qs=interesting_packages_qs,
+            logger=logger,
+        )
+
+    if logger:
+        logger(f"Successfully removed {ghost_package_count:,d} ghost Packages")
+
+
+def remove_ghost_package(package, interesting_packages_qs, logger=None):
+    if not package["type"] in RANGE_CLASS_BY_SCHEMES:
+        return 0
+
+    versionless_purl = PackageURL(**package)
+    purl_type = package["type"]
+    version_class = RANGE_CLASS_BY_SCHEMES[purl_type].version_class
+
+    try:
+        known_versions = [version_class(v.value) for v in versions(str(versionless_purl))]
+    except Exception as e:
+        if logger:
+            logger(f"An error occurred: {e}", level=logging.ERROR)
+        return 0
+
+    package_versions = interesting_packages_qs.filter(**package)
+    removed_packages = 0
+    for pkg in package_versions:
+        if version_class(pkg.version) not in known_versions:
+            pkg.delete()
```

Comment:
Also should we not instead remove the relationship? And leave the dangling package record and flag it as a "ghost"? This could be an improvement for the future? There could be cases where the package version may have been yanked from the package repository and would no longer be seen as a version there, yet it existed and was a vulnerable version. Here is a practical example:

https://registry.npmjs.org/gxm-reference-web-auth-server is a malicious package
See
- https://github.com/advisories/GHSA-6c6p-8v5p-j959
- https://www.npmjs.com/package/gxm-reference-web-auth-server

There is only now one version 0.1 left. But based on https://github.com/Odyhibit/NPMspy/tree/caa284392f88b4f189da1f06471ebf36880d78bc/tests/set1_malware_packages/gxm-reference-web-auth-server there were likely other versions back then. If we delete all versions except for 0.1 then we have a serious issue

See also other cases https://github.com/Odyhibit/NPMspy/


So to recap: two cases are problematic:

1. yanked package versions
2. forever malicious where all versions are problematic even if we cannot find them

Reply:
@pombredanne i'v added the status filed to package to flag malicious, ghost, yanked, valid and unknown. we are not deleting/removing package for now.